### ISSUE TITLE
Docked FAB gets clipped when keyboard is shown

### ISSUE THREAD

**Issue by moejay** [flutter/flutter#26100]
## Steps to Reproduce

1. run app ( `main.dart` below )
2. open keyboard ( fab onPressed will do it )

Actual: FAB clipped ( See end of issue for snippet to make it work )
<img width="150" alt="clipped-docked-fab" src="https://user-images.githubusercontent.com/1688957/50718173-da796280-105a-11e9-9c08-8c5886f7adac.png">
Expected: FAB not clipped
<img width="150" alt="non-clipped-docked-fab" src="https://user-images.githubusercontent.com/1688957/50718174-df3e1680-105a-11e9-8ca1-bbd45e860f48.png">

```dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

void showKeyboard () => SystemChannels.textInput.invokeMethod('TextInput.show');
void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      home: FabExample (),
    );
  }
}
class FabExample extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('FAB Test'),
      ),
      body: Center(),
      // FAB floats properly above the keyboard ( including suggestions )
//      floatingActionButtonLocation: FloatingActionButtonLocation.centerFloat,
      // FAB docked above the keyboard, not taking into account the full height
      floatingActionButtonLocation: FloatingActionButtonLocation.centerDocked,
      floatingActionButton: FloatingActionButton(
        onPressed: showKeyboard,
        child: Icon(Icons.keyboard),
      ),
    );
  }
}
```
## Logs

[flutter-run-verbose.log](https://github.com/flutter/flutter/files/2728979/flutter-run-verbose.log)

```
flutter-analyze
Analyzing docker_fab_example...
No issues found! (ran in 2.0s)
```

```
[√] Flutter (Channel stable, v1.0.0, on Microsoft Windows [Version 10.0.17134.472], locale en-US)
    • Flutter version 1.0.0 at d:\Work\SDK\flutter
    • Framework revision 5391447fae (5 weeks ago), 2018-11-29 19:41:26 -0800
    • Engine revision 7375a0f414
    • Dart version 2.1.0 (build 2.1.0-dev.9.4 f9ebf21297)

[√] Android toolchain - develop for Android devices (Android SDK 28.0.3)
    • Android SDK at D:\Work\SDK\android-sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-27, build-tools 28.0.3
    • ANDROID_HOME = D:\Work\SDK\android-sdk
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1136-b06)
    • All Android licenses accepted.

[√] Android Studio (version 3.2)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin version 31.3.1
    • Dart plugin version 181.5656
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1136-b06)

[√] Connected device (1 available)
    • Android SDK built for x86 • emulator-5554 • android-x86 • Android 8.0.0 (API 26) (emulator)

• No issues found!

```

### What made it work

Although am not entirely sure that's the correct solution, I was able to make it work by removing the division by `2` that happens on `fabY` in `_DockedFloatingActionButtonLocation` in `floating_action_button_location.dart`

```dart
import 'dart:math';

class FixedCenterDockedFabLocation extends FloatingActionButtonLocation {
  const FixedCenterDockedFabLocation();

  @protected
  double getDockedY(ScaffoldPrelayoutGeometry scaffoldGeometry) {
    final double contentBottom = scaffoldGeometry.contentBottom;
    final double bottomSheetHeight = scaffoldGeometry.bottomSheetSize.height;
    final double fabHeight = scaffoldGeometry.floatingActionButtonSize.height;
    final double snackBarHeight = scaffoldGeometry.snackBarSize.height;

    // In _DockedFloatingActionButtonLocation this value is
    // double fabY = contentBottom - fabHeight / 2.0;
    // ----------------------------------
    // VVVVVVVVVVVVVVVVVVVVVV
    double fabY = contentBottom - fabHeight;

    // The FAB should sit with a margin between it and the snack bar.
    if (snackBarHeight > 0.0)
      fabY = min(fabY, contentBottom - snackBarHeight - fabHeight - kFloatingActionButtonMargin);
    // The FAB should sit with its center in front of the top of the bottom sheet.
    if (bottomSheetHeight > 0.0)
      fabY = min(fabY, contentBottom - bottomSheetHeight - fabHeight / 2.0);

    final double maxFabY = scaffoldGeometry.scaffoldSize.height - fabHeight;
    return min(maxFabY, fabY);
  }
  @override
  Offset getOffset(ScaffoldPrelayoutGeometry scaffoldGeometry) {
    final double fabX = (scaffoldGeometry.scaffoldSize.width - scaffoldGeometry.floatingActionButtonSize.width) / 2.0;
    return Offset(fabX, getDockedY(scaffoldGeometry));
  }
}
```

**Comment by vvasc** [comment]
the correct approach to solve that, seems to me, something like:

```
 double bottomDistance = MediaQuery.of(context).viewInsets.bottom;
 double fabY = contentBottom + bottomDistance - fabHeight / 2.0;
```

where you have the size of keyboard on calculation of fabY

**Comment by flbaue** [comment]
Thanks for creating this issue, I am struggling with it as well.
But I think the proposed solutions are not correct yet.

It seems to me, that the ```contentBottom``` is always correct. Whether there is a keyboard or not. The problem is, that in case of the ```AppBottomBar``` we want the FAB to be lowered by half, hence the ```fabHeight / 2```, but incase there is a keyboard, it is different.

If there is a keyboard, the FAB should be either fully on top of the keyboard, or completely covered by it. (I am not sure which one is actually desired).

My temporary workaround is to check if there is a bottom inset. 
```final hasBottomInset = scaffoldGeometry.minInsets.bottom > 0;```
And if that is true, I can handle the fabHeight differently. That works fine for the keyboard case, but I don't know if there might be other cases.

**Comment by nickcheek** [comment]
This seems to make it work
```dart
double getDockedY(ScaffoldPrelayoutGeometry scaffoldGeometry) {
    final double contentBottom = scaffoldGeometry.contentBottom;
    final double bottomSheetHeight = scaffoldGeometry.bottomSheetSize.height;
    final double fabHeight = scaffoldGeometry.floatingActionButtonSize.height;
    final double snackBarHeight = scaffoldGeometry.snackBarSize.height;
    //added to see if the bottom distance is more than 0
    final double bottomDistance = scaffoldGeometry.minInsets.bottom;

    double fabY = contentBottom - fabHeight / 2.0 ;

    //This seems to make it go up enough to be fully visible.
    if (bottomDistance > 0.0)
      fabY = contentBottom - (fabHeight + fabHeight) / 2.0 ;
    // The FAB should sit with a margin between it and the snack bar.
    if (snackBarHeight > 0.0)
      fabY = math.min(fabY, contentBottom - snackBarHeight - fabHeight - kFloatingActionButtonMargin);
    // The FAB should sit with its center in front of the top of the bottom sheet.
    if (bottomSheetHeight > 0.0)
      fabY = math.min(fabY, contentBottom - bottomSheetHeight - fabHeight / 2.0);
```

**Comment by cassioseffrin** [comment]
There is an alternative, set on the parent Scafold the param:
 resizeToAvoidBottomPadding: false

**Comment by Bo177** [comment]
@cassioseffrin Thanks for sharing the second alternative. 
There's only one problem with it, you can still tap the FAB even when it's invisible. So instead of changing the opacity you could simply return null or the button:
`return keyboardIsOpened
  ? null
  : FloatingActionButton( 
    ...
  );`

**Comment by cassioseffrin** [comment]
@Bo177  indeed!  I am just returning a Container() instead null. And I am using _MediaQuery.of(context).viewInsets.bottom == 0_ to check if keyboard is shown

```
MediaQuery.of(context).viewInsets.bottom == 0 ? FloatingActionButton(
        onPressed: () {
          Navigator.push(
            context,
            MaterialPageRoute(
              builder: (context) => YourScreen(),
            ),
          );
        },
        child: Icon(
          Icons.add,
        ),
        elevation: 2.0,
        backgroundColor: Theme.of(context).primaryColor,
      ) : Container()
```

**Comment by TahaTesser** [comment]
<details> 
<summary>flutter doctor -v</summary> 

``` 
[✓] Flutter (Channel dev, 1.20.0-0.0.pre, on Mac OS X 10.15.5 19F101, locale
    en-GB)
    • Flutter version 1.20.0-0.0.pre at /Users/tahatesser/Code/flutter_dev
    • Framework revision d9653445f4 (8 days ago), 2020-06-09 18:43:03 -0400
    • Engine revision e8c13aa012
    • Dart version 2.9.0 (build 2.9.0-14.0.dev 5c1376615e)

 
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.0)
    • Android SDK at /Users/tahatesser/Code/sdk
    • Platform android-30, build-tools 30.0.0
    • ANDROID_HOME = /Users/tahatesser/Code/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.5)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.5, Build version 11E608c
    • CocoaPods version 1.9.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 46.0.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.46.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.11.0

[✓] Connected device (4 available)
    • SM M305F   • 32003c30dc19668f • android-arm64  • Android 10 (API 29)
    • macOS      • macOS            • darwin-x64     • Mac OS X 10.15.5 19F101
    • Web Server • web-server       • web-javascript • Flutter Tools
    • Chrome     • chrome           • web-javascript • Google Chrome
      83.0.4103.97

• No issues found!
``` 

</details>

**Comment by rodrigobl** [comment]
This was happening to me when using
 `Scaffold(
          floatingActionButtonLocation: FloatingActionButtonLocation.centerDocked,`

worked fine when I changed to

`Scaffold(
     floatingActionButtonLocation: FloatingActionButtonLocation.centerFloat,`

**Comment by RoyARG02** [comment]
I think I may have figured this out after digging through the rabbit hole.

## TL;DR

Here's the implementation:

```dart
double getOffsetY( ScaffoldPrelayoutGeometry scaffoldGeometry, double adjustment) {
    final double contentBottom = scaffoldGeometry.contentBottom;
    final double contentMargin = scaffoldGeometry.scaffoldSize.height - contentBottom;
    final double bottomViewPadding = scaffoldGeometry.minViewPadding.bottom;
    final double bottomSheetHeight = scaffoldGeometry.bottomSheetSize.height;
    final double fabHeight = scaffoldGeometry.floatingActionButtonSize.height;
    final double snackBarHeight = scaffoldGeometry.snackBarSize.height;
    final double bottomMinInset = scaffoldGeometry.minInsets.bottom;

    final double safeMargin = bottomMinInset + fabHeight / 2.0 < contentMargin
        ? 0.0
        : bottomMinInset == 0.0
            ? bottomViewPadding
            : fabHeight / 2.0 + kFloatingActionButtonMargin;

    double fabY = contentBottom - fabHeight / 2.0 - safeMargin;
    // The FAB should sit with a margin between it and the snack bar.
    if (snackBarHeight > 0.0)
      fabY = math.min( fabY, contentBottom - snackBarHeight - fabHeight - kFloatingActionButtonMargin);
    // The FAB should sit with its center in front of the top of the bottom sheet.
    if (bottomSheetHeight > 0.0)
      fabY = math.min(fabY, contentBottom - bottomSheetHeight - fabHeight / 2.0);
    final double maxFabY = scaffoldGeometry.scaffoldSize.height - fabHeight - safeMargin;
    return math.min(maxFabY, fabY);
  }
```

## Explanation

*This explanation is all I could figure out after an afternoon of tweaking. Feel free to correct any misunderstandings.*

The important properties for this implementation are explained below:

- **contentBottom**: The distance from the top of the `Scaffold` to the bottom of `Scaffold.body`.
- **contentMargin**: The difference between the entire height of the `Scaffold` and `contentBottom`. That leaves the space taken up by `Scaffold.bottomNavigationBar` or system keyboard. The important thing to note is that `contentMargin` will assume the value `max(Scaffold.bottomNavigationBar, <height of system keyboard>)`, so we do not have a way to detect the presence of keyboard using `contentMargin` alone. That is why I use...
- **bottomMinInset**: The amount of space taken up by system UI including keyboards.
- **bottomViewPadding**: This is a tricky one. PR #60248 which introduces this provides some context, but the real issue is that it *only applies in iOS*. Since iOS allows the `Scaffold` to paint behind the gesture navigation bar, this property adds an additional padding of `34 pixels` to make sure the `FloatingActionButton` stays interactive. This isn't the deal with Android however, so it has a value of `0.0` over there.

### How it works

What makes it work is the change of the value of `safeMargin`.
The earlier implementation

```dart
final double safeMargin = bottomViewPadding > contentMargin ? bottomViewPadding : 0.0;
```

Only accounted for the keyboard closed. The pseudo code can be summarized as follows:

```txt
if contentMargin > bottomMinInset + fabHeight / 2   (Scaffold.bottomNavigationBar is bigger in height than the keyboard, 
                                                    there is ample space to display a docked FAB)
  then safeMargin = 0.0
else if keyboard is closed
  then safeMargin = bottomViewPadding       (= 34.0px in iOS, 0.0 in Android)
else
  safeMargin = fabHeight / 2 + kFloatingActionButtonMargin  (this makes the docked FAB to behave like a 
                                                            floating FAB above the keyboard)
```

## Demo

Here's the fix in action (using `device_preview` package):

|iOS|Android|
|---|-------|
|![ios](https://user-images.githubusercontent.com/44899587/110336397-d5edd180-804a-11eb-9986-5f868e6a51fe.gif)|![android](https://user-images.githubusercontent.com/44899587/110336426-dbe3b280-804a-11eb-8d22-b498d9e85720.gif)|

The "disappearing and reappearing of the FAB" doesn't happen in my physical Android device, so it is possible this has something to do with the `device_preview` package.

The code used for the above demo is in [this](https://gist.github.com/RoyARG02/0aab15901b5ca55142dca41d80be6a31) gist.

**Comment by sils** [comment]
For future reference, this is an implementation following this idea minus a few quirks, with less lines of code and using more upstream code, i.e. hopefully better to maintain:

```
class FixedCenterDockedFabLocation extends StandardFabLocation
    with FabCenterOffsetX, FabDockedOffsetY {
  const FixedCenterDockedFabLocation();

  @override
  String toString() => 'FloatingActionButtonLocation.fixedCenterDocked';

  @override
  double getOffsetY(
      ScaffoldPrelayoutGeometry scaffoldGeometry, double adjustment) {
    final double bottomMinInset = scaffoldGeometry.minInsets.bottom;
    if (bottomMinInset > 0) {
      // Hide if there's a keyboard
      return 0;
    }
    return super.getOffsetY(scaffoldGeometry, adjustment);
  }
}
```

Feel free to use this code without limitations or attribution.

Also note that when using this, use it with `const`.
